### DISCUSSION TITLE
How come setting `Data` in a `State` turns this app unreactive?

### DISCUSSION THREAD

**Discussion by dirtyhenry** [pointfreeco/swift-composable-architecture#2308] [Q&A] [answered]
👋 I am probably missing something silly here but in a very little sample app, as soon as I add `Data` to my state, my app becomes unreactive, cf. demo 👇 

Is it a bad idea to store binary data in the state for some reason? 🤔 

[Full code here](https://github.com/dirtyhenry/camera-tca), but the gist of what is going on is here:

```swift
struct RootFeature: ReducerProtocol {
    struct State {
        var latestPhotoData: Data?
        var count: Int
    }

    enum Action {
        case incrementButtonTapped
        case loadDummyButtonTapped
    }

    var body: some ReducerProtocolOf<Self> {
        Reduce<State, Action> { state, action in
            switch action {
            case .incrementButtonTapped:
                state.count += 1
                return .none

            case .loadDummyButtonTapped:
                guard let image = UIImage(named: "DummyImage"),
                      let imageData = image.jpegData(compressionQuality: 1)
                else {
                    fatalError("No image found.")
                }
                state.latestPhotoData = imageData
                print("Set data")
                return .none
            }
        }
    }
}

extension RootFeature.State: Equatable {}
extension RootFeature.Action: Equatable {}
```

Thanks 🙏 

https://github.com/pointfreeco/swift-composable-architecture/assets/456506/c3a46d84-cb91-419d-b943-9b802bc7c87f

**Comment by stephencelis** [comment]
@dirtyhenry I'm not able to reproduce that using your project in Xcode 15 beta 4 (the view is completely replaced with "PHOTO"). Do you have more instructions on how to repro?

**Reply by dirtyhenry** [reply]
Thanks @stephencelis. I am still using Xcode 14.3.1 on a M2 Mac.

I restarted my laptop, and tested on both a simulator and a real device and I still have the same behavior as in the video above. This is getting bad for my hair loss. 🤔 😬

**Reply by dirtyhenry** [reply]
I also tested on Xcode 14.3 on an Intel Mac and the app still behaves the same as in the video.

**Reply by mbrandonw** [reply, marked as answer]
Hi @dirtyhenry, I was able to reproduce this in iOS 16, though it does seem fixed in iOS 17 somehow.

I can't explain it, but something about `Data` and `WithViewStore` makes SwiftUI get into a very weird state. You can avoid `WithViewState` by using `@ObservedObject` directly:

```swift
struct RootView: View {
  let store: StoreOf<RootFeature>
  @ObservedObject var viewStore: ViewStoreOf<RootFeature>
  init(store: StoreOf<RootFeature>) {
    self.store = store
    self.viewStore = ViewStore(self.store, observe: { $0 })
  }

  …
}
```

That fixes the problem somehow.

Additionally I would say that generally it is not necessary to store `Data` in state.  What is the reason to store the data in state? Is it needed to execute actual logic in your feature? Or is it just being loaded so that it can be used in the view? Most likely all of that can just go in the view, no need to involve the reducer.

**Reply by dirtyhenry** [reply]
> Is it needed to execute actual logic in your feature? Or is it just being loaded so that it can be used in the view?

I am working on a feature with a photo upload service. I guess I could:

* keep the photo in the view to display the photo thumbnail 
* respond to the action that receives the picture from the camera by returning the `EffectTask` in charge of the upload so that storing the photo data in the state is not necessary. I was anticipating a retry feature when I assumed i needed to store the photo in the state.

Anyways, thanks a lof for your help: i have a clearer path to implement this feature now 🙏
